hsmd/hsmd.c

```diff
@@ -151,74 +243,337 @@ static struct client *new_client(struct daemon_conn *master,
 	return c;
 }
 
-/**
- * hsm_peer_secret_base -- Derive the base secret seed for per-peer seeds
- *
- * This secret is shared by all channels/peers for the client. The
- * per-peer seeds will be generated from it by mixing in the
- * channel_id and the peer node_id.
- */
-static void hsm_peer_secret_base(struct secret *peer_seed_base)
+/* This is the common pattern for the tail of each handler. */
+static struct io_plan *req_reply(struct io_conn *conn,
+				 struct client *c,
+				 const u8 *msg_out TAKES)
+{
+	/*~ Write this out, then read the next one.  This works perfectly for
+	 * a simple request/response system like this, as we don't do any real
+	 * buffering: if they don't read from their end, eventually we won't
+	 * read any more commands from them once the pipe fills. */
+	return io_write_wire(conn, msg_out, client_read_next, c);
+}
```

Comment:
I am still confused about this comment.

So for every (complete/entire?) request message that is read, it sends a `reply` to the client to indicate it got it. But the client _may_ not read the `reply` from their end and thus that pipe (kernel buffer?) fills. And I guess the (`io_plan`) behaviour is that it stops reading request messages when the reply pipe is full?

Reply:
In terms of threads, this effectively means something like a large loop approximately like this:

    while (true) {
            msg_out = client_read_next(c, conn);
            io_write_wire(conn, msg_out);
    }

`io_write_wire` will "block" (i.e. not call its callback `client_read_next`) if the pipe it is writing to, is filled up.  So `client_read_next` will not run, and we will also stop reading the next command from the client until the client frees up space from the pipe (yes, kernel buffer it is).